app/gui2/src/util/ast/abstract.ts

```diff
@@ -0,0 +1,924 @@
+import { Token, Tree } from '@/generated/ast'
+import { parseEnso } from '@/util/ast'
+import { AstExtended } from '@/util/ast/extended.ts'
+import type { LazyObject } from '@/util/parserSupport'
+import * as random from 'lib0/random'
+import { reactive } from 'vue'
+import type { ExprId } from '../../../shared/yjsModel'
+import { IdMap } from '../../../shared/yjsModel'
+
+const committed = reactive(new Map<AstId, Ast>())
+/** New nodes, COW-copies of modified nodes, and pending deletions (nulls) */
+const edited = new Map<AstId, Ast | null>()
+
+/** Replace all committed values with the state of the uncommitted parse. */
+export function syncCommittedFromEdited() {
+  for (const id of committed.keys()) {
+    if (!edited.has(id)) committed.delete(id)
+  }
+  for (const [id, ast] of edited.entries()) {
+    if (ast === null) {
+      committed.delete(id)
+    } else {
+      committed.set(id, ast)
+    }
+  }
+}
+
+/** Returns a syntax node representing the current committed state of the given ID. */
+function getNode(id: AstId): Ast | null {
+  return committed.get(id) ?? null
+}
+
+/** Returns a syntax node representing the current state of the given ID, including any uncommitted modifications. */
+function getUncommitted(id: AstId): Ast | null {
+  if (edited.has(id)) {
+    return edited.get(id) ?? null
+  } else {
+    return committed.get(id) ?? null
+  }
+}
+
+declare const brandAstId: unique symbol
+declare const brandTokenId: unique symbol
+export type AstId = ExprId & { [brandAstId]: never }
+export type TokenId = ExprId & { [brandTokenId]: never }
+
+function newNodeId(): AstId {
+  return random.uuidv4() as AstId
+}
+function newTokenId(): TokenId {
+  return random.uuidv4() as TokenId
+}
+
+export type NodeChild = { whitespace?: string | undefined; node: AstId | Tok }
+type AstWithWhitespace = { whitespace?: string | undefined; node: AstId }
+type TokWithWhitespace = { whitespace?: string | undefined; node: Tok }
```

Comment:
What about making it parameterized? Like `NodeChild<T = AstId | Tok>` and replace `AstWithWhitespace`/`TokWithWhitespace` with `NodeChild<AstId>`/`NodeChild<Tok>`